sdk.go

```diff
@@ -96,27 +121,240 @@ const (
+// garbage collection, and memory must be manually deallocated by invoking
+// the Free method. Using instances of PluginEvents after invoking its Free()
+// method might lead to non-deterministic behavior.
+//
+// Here is an example of usage:
+//	func plugin_open(pState unsafe.Pointer, params *C.char, rc *int32) unsafe.Pointer {
+//		...
+//		// Create instance of sdk.PluginEvents
+//		pluginEvents, err := sdk.NewPluginEvents(maxNextBatchEvents, int64(sdk.MaxEvtSize))
+//		if err != nil {
+//			*rc = sdk.SSPluginFailure
+//			return nil
+//		}
+//
+//		// Store pluginEvents inside the plugin instance state
+//		is := &instanceState{
+//			...
+//			pluginEvents:       pluginEvents,
+//		}
+//		handle := state.NewStateContainer()
+//		state.SetContext(handle, unsafe.Pointer(is))
+//		*rc = sdk.SSPluginSuccess
+//		return handle
+//	}
+//
+//	func plugin_next(pState unsafe.Pointer, iState unsafe.Pointer, retEvt **C.ss_plugin_event) int32 {
+//		// Grab an instance of the event through sdk.PluginEvents
+//		is := (*instanceState)(state.Context(iState))
+//		event := is.pluginEvents.Get(0)
+//
+//		// Write the event data and set the timestamp
+//		eventData := "Sample event data... Hello World!"
+//		eventWriter := event.Writer()
+//		eventWriter.Write([]byte(eventData))
+//		event.SetTimestamp(uint64(time.Now().Unix()) * 1000000000)
+//
+//		// Set the result pointer to the internal event array buffer
+//		*retEvt = (*C.ss_plugin_event)(is.pluginEvents.ArrayPtr())
+//		return sdk.SSPluginSuccess
+//	}
+//
+type PluginEvents interface {
+	// Get returns an instance of sdk.PluginEvent at the eventIndex
+	// position inside the list.
+	Get(eventIndex int) PluginEvent
+	//
+	// Len returns the size of the list, namely the number of events
+	// it contains. Using Len coupled with Get allows iterating over
+	// all the events of the list.
+	Len() int
+	//
+	// Free takes care of de-allocating all the memory managed by
+	// instances of PluginEvents. Note that using the same instance
+	// after invoking its Free method might lead to non-determinitic
+	// behavior.
+	Free()
+	//
+	// ArrayPtr return an unsafe pointer to the underlying C array of
+	// ss_plugin_event. The returned pointer should only be used for
+	// read tasks or for being passed to the plugin framework.
+	// Writing in the memory pointed by this pointer is unsafe and might
+	// lead to non-deterministic behavior.
+	ArrayPtr() unsafe.Pointer
+}
+type pluginEvent struct {
+	data        BytesReadWriter
+	dataSize    int64
+	ssPluginEvt unsafe.Pointer
+}
+
+type pluginEvents []*pluginEvent
+
+// NewPluginEvents creates a new instance of sdk.PluginEvents.
+// The size argument indicates the length of the list, which is the amount
+// of events contained. Then dataSize argument indicates the maximum data
+// size of each event.
+func NewPluginEvents(size, dataSize int64) (PluginEvents, error) {
+	if size < 0 || size > MaxNextBatchEvents {
```

Comment:
```suggestion
	if size < 1 || size > MaxNextBatchEvents {
```
`PluginEvents` with 0 events is useless since there's no way to grow its size later.
Furthermore, forcing a size greater than 0 would allow us to avoid checking for len == 0 (see below).